Userland/Libraries/LibGUI/ZoomableView.cpp

```diff
@@ -0,0 +1,179 @@
+/*
+ * Copyright (c) 2022, Mustafa Quraish <mustafa@serenityos.org>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include "ZoomableView.h"
+
+namespace GUI {
+
+void ZoomableView::set_scale(float new_scale)
+{
+    if (m_original_rect.is_null())
+        return;
+
+    m_scale = new_scale;
+    Gfx::IntSize new_size;
+    new_size.set_width(m_original_rect.width() * m_scale);
+    new_size.set_height(m_original_rect.height() * m_scale);
+    m_image_rect.set_size(new_size);
+
+    if (on_scale_change)
+        on_scale_change(m_scale);
+
+    relayout();
+}
+
+void ZoomableView::scale_by(float delta)
+{
+    float new_scale = m_scale * AK::exp2(delta);
+    set_scale(new_scale);
+}
+
+void ZoomableView::scale_centered(float new_scale, Gfx::IntPoint const& center)
+{
+    if (m_original_rect.is_null())
+        return;
+
+    new_scale = max(min(new_scale, m_max_scale), m_min_scale);
+
+    if (new_scale == m_scale)
+        return;
+
+    Gfx::FloatPoint focus_point {
+        center.x() - width() / 2.0f,
+        center.y() - height() / 2.0f
+    };
+    m_origin = (m_origin + focus_point) * (new_scale / m_scale) - focus_point;
+    set_scale(new_scale);
+}
+
+void ZoomableView::start_panning(Gfx::IntPoint const& position)
+{
+    m_saved_cursor = override_cursor();
+    set_override_cursor(Gfx::StandardCursor::Drag);
+    m_pan_start = m_origin;
+    m_pan_mouse_pos = position;
+    m_is_panning = true;
+}
+
+void ZoomableView::stop_panning()
+{
+    m_is_panning = false;
+    set_override_cursor(m_saved_cursor);
+}
+
+void ZoomableView::pan_to(Gfx::IntPoint const& position)
```

Comment:
I'd appreciate a comment here explaining in what coordinate system `position` is.

Reply:
Good point! Added one, happy to add more if anything's unclear 😅